CONTRACTS: Moving `check_frame_conditions` to `instrument_assigns_clauset`.

remi-delmas-3000 opened the pull request:
To further encapsulate the the assigns clause checking algorithm in `instrument_assigns_clause.cpp`

-  the `check_frame_conditions` method and related functions have been moved from `contracts.cpp` and `utils.cpp` to `instrument_assigns_clause.cpp`.
- the former `track_static_locals` method is now available in two different versions:
  - `track_static_locals(dest)` which tracks all local static variables touched directly or indirectly by the whole instrumented function (for function contracts)
  - `track_static_locals_from_calls(begin, end, dest)` which tracks all local static variables declared by functions that are called between the begin and end instructions (for loop contracts).

No functionality or performance impact expected.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -257,12 +257,11 @@ void code_contractst::check_apply_loop_contracts(
   // assigns clause snapshots
   goto_programt snapshot_instructions;
 
+  // track static locals
+  // TODO: restrict to variables declared in the loop's scope
+  // or the functions called in the loop body
```
Are we tracking this somewhere (e..g, GitHub issue)?

remi-delmas-3000 replied:
yes we are tracking it

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -27,8 +27,62 @@ Date: January 2022
 #include "utils.h"
 
-///// PUBLIC METHODS /////
+void add_pragma_disable_pointer_checks(source_locationt &location)
+{
+  location.add_pragma("disable:pointer-check");
+  location.add_pragma("disable:pointer-primitive-check");
+  location.add_pragma("disable:pointer-overflow-check");
+  location.add_pragma("disable:signed-overflow-check");
+  location.add_pragma("disable:unsigned-overflow-check");
+  location.add_pragma("disable:conversion-check");
+}
```
What is the motivation to move this away from `utils` and add it in here? Perhaps it's only applied to assigns clauses today, but there might be value in using it in other places in the future.

feliperodri replied:
Same for other `pragma` instrumentation related changes.

remi-delmas-3000 replied:
all these pragmas are solely used in assigns clause checking, no need to expose them in utils.

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -27,8 +27,62 @@ Date: January 2022
+  return prog;
+}
+
+/// a local pragma used to keep track and skip already instrumented instructions
+#define CONTRACT_PRAGMA_DISABLE_ASSIGNS_CHECK "contracts:disable:assigns-check"
```
```
/// A local pragma used to keep track and skip already instrumented instructions.
```

I'd also place all definitions right after includes.

remi-delmas-3000 replied:
made it a `const std::string` placed at the top

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -27,8 +27,62 @@ Date: January 2022
+  return prog;
+}
+
+///// PUBLIC METHODS /////
```
I think these comments are unnecessary.

remi-delmas-3000 replied:
removed

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -102,13 +158,14 @@ void instrument_spec_assignst::track_static_locals(goto_programt &dest)
         // - a static lifetime
         // - non empty location function
         // - this function appears in the call graph of the function
+        // - it is accepted by the predicate
```
which predicate?

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -118,6 +175,26 @@ void instrument_spec_assignst::track_static_locals(goto_programt &dest)
+  collect_static_locals_from_root_function(function_id, dest);
+}
+
+void instrument_spec_assignst::track_static_locals_from_calls(
```
Why don't we cover this functions? (see Codecov reports).

remi-delmas-3000 replied:
removed

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.h:
```diff
@@ -131,6 +131,51 @@ class car_exprt : public exprt
   }
 };
 
+/// \brief Adds a pragma on a source location disable all pointer checks.
+///
+/// The disabled checks are: "pointer-check", "pointer-primitive-check",
+/// "pointer-overflow-check", "signed-overflow-check",
+//  "unsigned-overflow-check", "conversion-check".
+void add_pragma_disable_pointer_checks(source_locationt &source_location);
+
+/// \brief Adds pragmas on a GOTO instruction to disable all pointer checks.
+///
+/// \param instr: A mutable reference to the GOTO instruction.
+/// \return The same reference after mutation (i.e., adding the pragma).
+goto_programt::instructiont &
+add_pragma_disable_pointer_checks(goto_programt::instructiont &instr);
+
+/// \brief Adds pragmas on all instructions in a GOTO program
+/// to disable all pointer checks.
+///
+/// \param prog: A mutable reference to the GOTO program.
+/// \return The same reference after mutation (i.e., adding the pragmas).
+goto_programt &add_pragma_disable_pointer_checks(goto_programt &prog);
+
+/// \brief Adds a pragma on a source_locationt to disable inclusion checking.
+void add_pragma_disable_assigns_check(source_locationt &source_location);
+
+/// \brief Adds a pragma on a GOTO instruction to disable inclusion checking.
+///
+/// \param instr: A mutable reference to the GOTO instruction.
+/// \return The same reference after mutation (i.e., adding the pragma).
+goto_programt::instructiont &
+add_pragma_disable_assigns_check(goto_programt::instructiont &instr);
+
+/// \brief Adds pragmas on all instructions in a GOTO program
+///        to disable inclusion checking on them.
+///
+/// \param prog: A mutable reference to the GOTO program.
+/// \return The same reference after mutation (i.e., adding the pragmas).
+goto_programt &add_pragma_disable_assigns_check(goto_programt &prog);
```
I'd keep all these functions in `utils` module.

remi-delmas-3000 replied:
they used to be in the utils module because the assigns clause checking logic was spread between several files, now that everything is regrouped in a single place these functions only need to live in that place.

feliperodri reviewed (commented):
Only a few clarifications before approval.

feliperodri commented:
>     * the `check_frame_conditions` method and related functions have been moved from `contracts.cpp` and `utils.cpp` to `instrument_assigns_clause.cpp`.
> 
>     * the former `track_static_locals` method is now available in two different versions:
>       
>       * `track_static_locals_whole_function(dest)` which tracks all local static variables touched directly or indirectly by the whole instrumented function (for function contracts)
>       * `track_static_locals_in_range(begin, end, dest)` which tracks all local static variables touched by function calls found between the begin and end instructions (for loop contracts).

@remi-delmas-3000 I'd create one commit for each one of these two main bullet points.

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -27,8 +27,62 @@ Date: January 2022
+  location.add_pragma("disable:conversion-check");
+}
+
+goto_programt::instructiont &
+add_pragma_disable_pointer_checks(goto_programt::instructiont &instr)
+{
+  add_pragma_disable_pointer_checks(instr.source_location_nonconst());
+  return instr;
```
The coverage annotation suggests that this code may be unnecessary?

remi-delmas-3000 replied:
indeed, not needed anymore

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -27,8 +27,62 @@ Date: January 2022
+  return instr;
+}
 
+goto_programt &add_pragma_disable_pointer_checks(goto_programt &prog)
+{
+  Forall_goto_program_instructions(it, prog)
+    add_pragma_disable_pointer_checks(*it);
+  return prog;
```
This one might also be unnecessary?

remi-delmas-3000 replied:
yes

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -513,3 +682,117 @@ void instrument_spec_assignst::invalidate_heap_and_spec_aliases(
+  {
+    if(
+      skip_function_params == skip_function_paramst::No &&
+      ns.lookup(symbol_expr->get_identifier()).is_parameter)
+      return true;
+
```
Nit pick: braces around the body of an if statement with a multi-line decision, please.

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -513,3 +682,117 @@ void instrument_spec_assignst::invalidate_heap_and_spec_aliases(
+  const goto_programt::const_targett &target,
+  const optionalt<cfg_infot> &cfg_info_opt) const
+{
+  if(cfg_info_opt.has_value())
+    return cfg_info_opt.value().is_not_local_or_dirty_local(
+      target->decl_symbol().get_identifier());
```
Braces around multi-line body, please.

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -513,3 +682,117 @@ void instrument_spec_assignst::invalidate_heap_and_spec_aliases(
+  goto_programt &body,
+  optionalt<cfg_infot> &cfg_info_opt)
+{
+  INVARIANT(
+    instruction_it->is_function_call(),
+    "The first argument of instrument_call_statement should always be "
+    "a function call");
```
`PRECONDITION` may be more specific than the most generic `INVARIANT` here.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.h:
```diff
@@ -131,6 +131,51 @@ class car_exprt : public exprt
+/// Skip or do not skip assignments to function parameters
+enum class skip_function_paramst
+{
+  Yes,
+  No
```
All uppercase, please.

remi-delmas-3000 replied:
fixed

Commit:
CONTRACTS: Move check_frame_conditions to instrument_assigns_clauset

The `check_frame_conditions` method and related methods
have been moved from `contracts.cpp` and `utils.cpp`
to `instrument_assigns_clause.cpp`.

No functionality or performance impact expected.

remi-delmas-3000 commented:
I addressed all comments, removed dead code, and removed the changes to local static detection. I think it is stable now.
